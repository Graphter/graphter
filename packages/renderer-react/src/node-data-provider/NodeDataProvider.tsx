import React, { createContext, useContext } from 'react';
import { NodeConfig } from "microo-core";

interface DataProviderProps {
  nodeDataHook: NodeDataHook,
  config: NodeConfig,
  children: any
}

export interface NodeDataHook {
  (
    config: NodeConfig,
    storeId: string,
    originalNodeData: any
  ): [ any, (value: any) => void ]
}

const Context = createContext<{ nodeDataHook: NodeDataHook, config: NodeConfig } | null>(null);

export function useNodeData(storeId: string, jsonPointer: string, originalNodeData: any): [ any, (value: any) => void ] {
  const ctx = useContext(Context);
  if (!ctx || !ctx.nodeDataHook) throw new Error(`Couldn't find a NodeDataHook to use.`);
  return ctx.nodeDataHook(ctx.config, storeId, originalNodeData);
}

export default function NodeDataProvider(
  {
    config,
    nodeDataHook,
    children
  }: DataProviderProps
) {
  return (
    <Context.Provider value={{ nodeDataHook, config }}>
      {children}
    </Context.Provider>
  );
}