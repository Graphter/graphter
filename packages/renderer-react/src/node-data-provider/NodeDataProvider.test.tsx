import React from 'react';
import { render } from '@testing-library/react';

import NodeDataProvider, { useNodeData, useArrayNodeData } from './NodeDataProvider'

function MockDataConsumer(props: any) {
  useNodeData(
    props.path,
    props.config,
    props.originalNodeData,
    props.committed
  )
  return null
}
function MockArrayDataConsumer<D>(props: any) {
  useArrayNodeData(
    props.path,
    props.config,
    props.originalChildData,
    props.committed
  )
  return null
}

describe('<NodeDataProvider />', () => {

  it('should render children', () => {
    const { queryByTestId } = render(
      <NodeDataProvider
        instanceId={'some-instance-id'}
        nodeDataHook={jest.fn()}
        arrayNodeDataHook={jest.fn()} >
        <div data-testid='some-child-component'>Some child component</div>
      </NodeDataProvider>
    )
    expect(queryByTestId('some-child-component')).not.toBeNull();
  })

  describe('when using useNodeData', () => {
    it('should use the hook supplied to the provider to retrieve data', () => {
      const nodeDataHookMock = jest.fn().mockReturnValue([ ])
      render(
        <NodeDataProvider
          instanceId={'some-instance-id'}
          nodeDataHook={nodeDataHookMock}
          arrayNodeDataHook={jest.fn()} >
          <MockDataConsumer
            path={['/']}
            config={{
              id: 'name',
              type: 'string'
            }}
            originalNodeData='The Name'
            committed={true}
          />
        </NodeDataProvider>
      )
      expect(nodeDataHookMock).toHaveBeenCalledWith(
        ['/'],
        {
          id: 'name',
          type: 'string'
        },
        'The Name',
        true
      )
    })
    it('should error if no provider is defined', () => {
      expect.assertions(1)
      try {
        render(
          <MockDataConsumer
            path={['/']}
            config={{
              id: 'name',
              type: 'string'
            }}
            originalNodeData='The Name'
            committed={true}
          />
        )
      } catch(err) {
        expect(err.message).toMatchSnapshot()
      }
    })
    it('should error if no hook is supplied to the provider', () => {
      expect.assertions(1)
      try {
        render(
          <NodeDataProvider
            instanceId={'some-instance-id'}
            // @ts-ignore
            nodeDataHook={null}
            arrayNodeDataHook={jest.fn()} >
            <MockDataConsumer
              path={['/']}
              config={{
                id: 'name',
                type: 'string'
              }}
              originalNodeData='The Name'
              committed={true}
            />
          </NodeDataProvider>
        )
      } catch(err) {
        expect(err.message).toMatchSnapshot()
      }
    })
  })
  describe('useArrayNodeData', () => {
    it('should use the hook supplied to the provider to retrieve data', () => {
      const arrayNodeDataHookMock = jest.fn().mockReturnValue({})
      render(
        <NodeDataProvider
          instanceId={'some-instance-id'}
          nodeDataHook={jest.fn()}
          arrayNodeDataHook={arrayNodeDataHookMock} >
          <MockArrayDataConsumer
            path={['/']}
            config={{
              id: 'tags',
              type: 'list'
            }}
            originalChildData={[ 'tagA', 'tagB' ]}
            committed={true}
          />
        </NodeDataProvider>
      )
      expect(arrayNodeDataHookMock).toHaveBeenCalledWith(
        ['/'],
        {
          id: 'tags',
          type: 'list'
        },
        [ 'tagA', 'tagB' ],
        true
      )
    })
    it('should error if no provider is defined', () => {
      expect.assertions(1)
      try {
        render(
          <MockArrayDataConsumer
            path={['/']}
            config={{
              id: 'tags',
              type: 'list'
            }}
            originalChildData={[ 'tagA', 'tagB' ]}
            committed={true}
          />
        )
      } catch(err) {
        expect(err.message).toMatchSnapshot()
      }
    })
    it('should error if no hook is supplied to the provider', () => {
      expect.assertions(1)
      try {
        render(
          <NodeDataProvider
            instanceId={'some-instance-id'}
            nodeDataHook={jest.fn()}
            // @ts-ignore
            arrayNodeDataHook={null} >
            <MockArrayDataConsumer
              path={['/']}
              config={{
                id: 'tags',
                type: 'list'
              }}
              originalChildData={[ 'tagA', 'tagB' ]}
              committed={true}
            />
          </NodeDataProvider>
        )
      } catch(err) {
        expect(err.message).toMatchSnapshot()
      }
    })
    it('should error if original data is not an array', () => {
      expect.assertions(1)
      try {
        render(
          <NodeDataProvider
            instanceId={'some-instance-id'}
            nodeDataHook={jest.fn()}
            arrayNodeDataHook={jest.fn()} >
            <MockArrayDataConsumer
              path={['/']}
              config={{
                id: 'tags',
                type: 'list'
              }}
              originalChildData={'string data'}
              committed={true}
            />
          </NodeDataProvider>
        )
      } catch(err) {
        expect(err.message).toMatchSnapshot()
      }
    })
  })
})
export {}