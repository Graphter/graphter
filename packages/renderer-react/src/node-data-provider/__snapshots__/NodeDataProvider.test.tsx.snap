// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NodeDataProvider /> useArrayNodeData should error if no hook is supplied to the provider 1`] = `"Couldn't find an ArrayNodeDataHook or context to use."`;

exports[`<NodeDataProvider /> useArrayNodeData should error if no provider is defined 1`] = `"Couldn't find an ArrayNodeDataHook or context to use."`;

exports[`<NodeDataProvider /> useArrayNodeData should error if original data is not an array 1`] = `"'list' renderer only works with arrays"`;

exports[`<NodeDataProvider /> when using useNodeData should error if no hook is supplied to the provider 1`] = `"Couldn't find a NodeDataHook or context to use."`;

exports[`<NodeDataProvider /> when using useNodeData should error if no provider is defined 1`] = `"Couldn't find a NodeDataHook or context to use."`;
